runners/core-java/src/main/java/org/apache/beam/runners/core/OutputAndTimeBoundedSplittableProcessElementInvoker.java

```diff
@@ -0,0 +1,231 @@
+import org.apache.beam.sdk.util.SideInputReader;
+import org.apache.beam.sdk.util.WindowedValue;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TupleTag;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A {@link SplittableProcessElementInvoker} that requests a checkpoint after the {@link
+ * DoFn.ProcessElement} call either outputs at least a given number of elements (in total over all
+ * outputs), or runs for the given duration.
+ */
+public class OutputAndTimeBoundedSplittableProcessElementInvoker<
+        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
+    extends SplittableProcessElementInvoker<InputT, OutputT, RestrictionT, TrackerT> {
+  private final DoFn<InputT, OutputT> fn;
+  private final PipelineOptions pipelineOptions;
+  private final OutputWindowedValue<OutputT> output;
+  private final SideInputReader sideInputReader;
+  private final ScheduledExecutorService executor;
+  private final int maxNumOutputs;
+  private final Duration maxDuration;
+
+  /**
+   * Creates a new invoker from components.
+   *
+   * @param fn The original {@link DoFn}.
+   * @param pipelineOptions {@link PipelineOptions} to include in the {@link DoFn.ProcessContext}.
+   * @param output Hook for outputting from the {@link DoFn.ProcessElement} method.
+   * @param sideInputReader Hook for accessing side inputs.
+   * @param executor Executor on which a checkpoint will be scheduled after the given duration.
+   * @param maxNumOutputs Maximum number of outputs, in total over all output tags, after which a
+   *     checkpoint will be requested. This is a best-effort request - the {@link DoFn} may output
+   *     more after receiving the request.
+   * @param maxDuration Maximum duration of the {@link DoFn.ProcessElement} call after which a
+   *     checkpoint will be requested. This is a best-effort request - the {@link DoFn} may run for
+   *     longer after receiving the request.
+   */
+  public OutputAndTimeBoundedSplittableProcessElementInvoker(
+      DoFn<InputT, OutputT> fn,
+      PipelineOptions pipelineOptions,
+      OutputWindowedValue<OutputT> output,
+      SideInputReader sideInputReader,
+      ScheduledExecutorService executor,
+      int maxNumOutputs,
+      Duration maxDuration) {
+    this.fn = fn;
+    this.pipelineOptions = pipelineOptions;
+    this.output = output;
+    this.sideInputReader = sideInputReader;
+    this.executor = executor;
+    this.maxNumOutputs = maxNumOutputs;
+    this.maxDuration = maxDuration;
+  }
+
+  @Override
+  public Result invokeProcessElement(
+      DoFnInvoker<InputT, OutputT> invoker,
+      final WindowedValue<InputT> element,
+      final TrackerT tracker) {
+    final ProcessContext processContext = new ProcessContext(element, tracker);
+    DoFn.ProcessContinuation cont =
+        invoker.invokeProcessElement(
+            new DoFnInvoker.FakeArgumentProvider<InputT, OutputT>() {
```

Comment:
Done, put in all the methods explicitly. A separate PR removes the annoying inputProvider/outputReceiver methods.